src/classes/RD_RecurringDonations.cls

```diff
@@ -795,77 +766,391 @@ public class RD_RecurringDonations {
+            , List<Opportunity> openOpportunities
+            , TDTM_Runnable.DmlWrapper dmlWrapper) {
+
+        npe03__Recurring_Donations_Settings__c rds = UTIL_CustomSettingsFacade.getRecurringDonationsSettings();
+
+        //build a map of recurring donation ID and opps
+        Map<Id, List<Opportunity>> RDOppMap = new Map<Id, List<Opportunity>>();
+        for (Opportunity o : openOpportunities){
+            if (RDOppMap.containsKey(o.npe03__Recurring_Donation__c)) {
+                List<Opportunity> opps = RDOppMap.get(o.npe03__Recurring_Donation__c);
+                opps.add(o);
+                RDOppMap.put(o.npe03__Recurring_Donation__c, opps);
             }
-            else{
-                set<id> rdset = new set<id>();
-                for (npe03__Recurring_Donation__c r : reinsertopps)
-                    rdset.add(r.id);
-                insertOppsOnRecurringDonationInsertFuture(rdset);   
+            else {
+                RDOppMap.put(o.npe03__Recurring_Donation__c, new List<Opportunity>{o});
+            }
+        }
+        // add to the map any RDs that came in with no Opportunities
+        for (npe03__Recurring_Donation__c rd : recurringDonations) {
+            if (!RDOppMap.containsKey(rd.Id)) {
+                RDOppMap.put(rd.Id, new List<Opportunity>());
+            }
+        }
+
+        List<Opportunity> updateOppsList = new List<Opportunity>();
+        List<Opportunity> closeOppsList = new List<Opportunity>();
+        List<Opportunity> insertOppsList = new List<Opportunity>();
+
+        integer forecast_months = (integer)rds.npe03__Opportunity_Forecast_Months__c;
+        Date dtForecast =  system.today().toStartOfMonth().addMonths(forecast_months);
+
+        for (npe03__Recurring_Donation__c r : recurringDonations) {
+
+            Date nextDate = getStartDate(r);
+
+            integer rdcounter = 1;
+            if (r.npe03__Total_Paid_Installments__c > 0) {
+                rdcounter = (integer)r.npe03__Total_Paid_Installments__c + 1;
+            }
+
+            // loop the existing opportunities
+            for (Opportunity o : RDOppMap.get(r.Id)) {
+                if (nextDate <= dtForecast) {
+                    // fix close dates on existing opportunities
+                    o.CloseDate = nextDate;
+
+                    // TODO: are these really all the fields necessary to sync to oppties on an update?
+                    // This is borrowed from roughly line 700
+                    o.Amount = r.npe03__Amount__c;
+                    if (r.npe03__Recurring_Donation_Campaign__c != null && rds.npe03__Add_Campaign_to_All_Opportunites__c) {
+                        o.CampaignId = r.npe03__Recurring_Donation_Campaign__c;
+                    }
+
+                    if (r.npe03__Organization__c != null) {
+                        o.AccountID = r.npe03__Organization__c;
+                    }
+
+                    //modify the name
+                    String oName = '';
```

Comment:
I've noticed there are a number of blocks of code in this class that concatenate some values into an Opportunity name. If feasible, it would be helpful to refactor these into a single method that build the Opportunity.Name value using parameters passed to the method (such as RecurringDonation__c, Opportunity, CloseDate, etc.)

Reply:
Refactored into utility.